### ISSUE TITLE
Unable to find eslint configuration file: eslint checker disabled

### ISSUE THREAD

**Issue by marc0s** [flycheck/flycheck#1155]
# Expected behavior

Have eslint checker enabled when enabling flycheck

# Current behavior

eslint checker is disabled because of missing configuration file *if emacs isn't launched from the path containing the configuration file*

# Environment

* GNU Emacs 25.1.1 (x86_64-pc-linux-gnu, GTK+ Version 3.20.9)
 of 2016-09-21
* Flycheck version: 31snapshot (package: 20161030.316)
* System:           x86_64-pc-linux-gnu (Debian testing)
* eslint v3.9.1
* Repo with reproducible code https://github.com/marc0s/flycheck-eslint-issue (attempting to `C-x f ~/path/to/index.js` raises the flycheck error buffer)

**Comment by Simplify** [comment]
Hi @marc0s,

What happens when you run `./home/marcos/.nvm/v4.6.1/bin/eslint --print-config .` in directory containing your `index.js` file? This is exactly what Flycheck does to check if `javascript-eslint` can be used.
I can see `--print-config` output in your repo, but just want to be sure that output is generated from the same ESlint that Emacs is using. Very often there is one eslint executable in project `node_modules` directory or one in old node installation that gets to be used in shell, but not in Emacs. 

Where and how are `standard` and `promise` plugins installed? If ESlint is installed globally, plugins needs to be installed that way too or command above will not exit with status code 0.

Also, your config file needs to be located where ESlint can find it. ESlint will use path to directory containing your js file and go down to your home directory. Flycheck no longer search for your ESlint config, not you can specify location of config file inside Flycheck. We just "ask" ESlint to tell us if we can use it for specific javascript file or not. If ESlint can't find config file, or detects problem in configuration (plugins needs to be installed same way as ESlint is installed) ESlint will exit with status code other than 0 and we will disable `javascript-eslint` for that buffer.

**Comment by marc0s** [comment]
Hi @Simplify 

Thanks for your detailed answer.

I'm aware of the need to use globally installed plugins, they already are.  Here the output you asked for:

```
~/tmp/flycheck-eslint-issue master*  
❯ /home/marcos/.nvm/v4.6.1/bin/eslint --print-config .
{
  "globals": {
    "Array": false,
    "ArrayBuffer": false,
    "Boolean": false,
    "constructor": false,
    "DataView": false,
    "Date": false,
    "decodeURI": false,
    "decodeURIComponent": false,
    "encodeURI": false,
    "encodeURIComponent": false,
    "Error": false,
    "escape": false,
    "eval": false,
    "EvalError": false,
    "Float32Array": false,
    "Float64Array": false,
    "Function": false,
    "hasOwnProperty": false,
    "Infinity": false,
    "Int16Array": false,
    "Int32Array": false,
    "Int8Array": false,
    "isFinite": false,
    "isNaN": false,
    "isPrototypeOf": false,
    "JSON": false,
    "Map": false,
    "Math": false,
    "NaN": false,
    "Number": false,
    "Object": false,
    "parseFloat": false,
    "parseInt": false,
    "Promise": false,
    "propertyIsEnumerable": false,
    "Proxy": false,
    "RangeError": false,
    "ReferenceError": false,
    "Reflect": false,
    "RegExp": false,
    "Set": false,
    "String": false,
    "Symbol": false,
    "SyntaxError": false,
    "System": false,
    "toLocaleString": false,
    "toString": false,
    "TypeError": false,
    "Uint16Array": false,
    "Uint32Array": false,
    "Uint8Array": false,
    "Uint8ClampedArray": false,
    "undefined": false,
    "unescape": false,
    "URIError": false,
    "valueOf": false,
    "WeakMap": false,
    "WeakSet": false,
    "__dirname": false,
    "__filename": false,
    "arguments": false,
    "Buffer": false,
    "clearImmediate": false,
    "clearInterval": false,
    "clearTimeout": false,
    "console": false,
    "exports": true,
    "GLOBAL": false,
    "global": false,
    "Intl": false,
    "module": false,
    "process": false,
    "require": false,
    "root": false,
    "setImmediate": false,
    "setInterval": false,
    "setTimeout": false,
    "document": false,
    "navigator": false,
    "window": false
  },
  "env": {
    "es6": true,
    "node": true
  },
  "rules": {
    "accessor-pairs": 2,
    "arrow-spacing": [
      2,
      {
        "before": true,
        "after": true
      }
    ],
    "block-spacing": [
      2,
      "always"
    ],
    "brace-style": [
      2,
      "1tbs",
      {
        "allowSingleLine": true
      }
    ],
    "camelcase": [
      2,
      {
        "properties": "never"
      }
    ],
    "comma-dangle": [
      2,
      "never"
    ],
    "comma-spacing": [
      2,
      {
        "before": false,
        "after": true
      }
    ],
    "comma-style": [
      2,
      "last"
    ],
    "constructor-super": 2,
    "curly": [
      2,
      "multi-line"
    ],
    "dot-location": [
      2,
      "property"
    ],
    "eol-last": 2,
    "eqeqeq": [
      2,
      "allow-null"
    ],
    "func-call-spacing": [
      2,
      "never"
    ],
    "handle-callback-err": [
      2,
      "^(err|error)$"
    ],
    "indent": [
      2,
      2,
      {
        "SwitchCase": 1
      }
    ],
    "key-spacing": [
      2,
      {
        "beforeColon": false,
        "afterColon": true
      }
    ],
    "keyword-spacing": [
      2,
      {
        "before": true,
        "after": true
      }
    ],
    "new-cap": [
      2,
      {
        "newIsCap": true,
        "capIsNew": false
      }
    ],
    "new-parens": 2,
    "no-array-constructor": 2,
    "no-caller": 2,
    "no-class-assign": 2,
    "no-cond-assign": 2,
    "no-const-assign": 2,
    "no-constant-condition": [
      2,
      {
        "checkLoops": false
      }
    ],
    "no-control-regex": 2,
    "no-debugger": 2,
    "no-delete-var": 2,
    "no-dupe-args": 2,
    "no-dupe-class-members": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 2,
    "no-duplicate-imports": 2,
    "no-empty-character-class": 2,
    "no-empty-pattern": 2,
    "no-eval": 2,
    "no-ex-assign": 2,
    "no-extend-native": 2,
    "no-extra-bind": 2,
    "no-extra-boolean-cast": 2,
    "no-extra-parens": [
      2,
      "functions"
    ],
    "no-fallthrough": 2,
    "no-floating-decimal": 2,
    "no-func-assign": 2,
    "no-global-assign": 2,
    "no-implied-eval": 2,
    "no-inner-declarations": [
      2,
      "functions"
    ],
    "no-invalid-regexp": 2,
    "no-irregular-whitespace": 2,
    "no-iterator": 2,
    "no-label-var": 2,
    "no-labels": [
      2,
      {
        "allowLoop": false,
        "allowSwitch": false
      }
    ],
    "no-lone-blocks": 2,
    "no-mixed-spaces-and-tabs": 2,
    "no-multi-spaces": 2,
    "no-multi-str": 2,
    "no-multiple-empty-lines": [
      2,
      {
        "max": 1
      }
    ],
    "no-native-reassign": 2,
    "no-negated-in-lhs": 2,
    "no-new": 2,
    "no-new-func": 2,
    "no-new-object": 2,
    "no-new-require": 2,
    "no-new-symbol": 2,
    "no-new-wrappers": 2,
    "no-obj-calls": 2,
    "no-octal": 2,
    "no-octal-escape": 2,
    "no-path-concat": 2,
    "no-proto": 2,
    "no-redeclare": 2,
    "no-regex-spaces": 2,
    "no-return-assign": [
      2,
      "except-parens"
    ],
    "no-self-assign": 2,
    "no-self-compare": 2,
    "no-sequences": 2,
    "no-shadow-restricted-names": 2,
    "no-sparse-arrays": 2,
    "no-tabs": 2,
    "no-template-curly-in-string": 2,
    "no-this-before-super": 2,
    "no-throw-literal": 2,
    "no-trailing-spaces": 2,
    "no-undef": 2,
    "no-undef-init": 2,
    "no-unexpected-multiline": 2,
    "no-unmodified-loop-condition": 2,
    "no-unneeded-ternary": [
      2,
      {
        "defaultAssignment": false
      }
    ],
    "no-unreachable": 2,
    "no-unsafe-finally": 2,
    "no-unsafe-negation": 2,
    "no-unused-vars": [
      2,
      {
        "vars": "all",
        "args": "none"
      }
    ],
    "no-useless-call": 2,
    "no-useless-computed-key": 2,
    "no-useless-constructor": 2,
    "no-useless-escape": 2,
    "no-useless-rename": 2,
    "no-whitespace-before-property": 2,
    "no-with": 2,
    "object-property-newline": [
      2,
      {
        "allowMultiplePropertiesPerLine": true
      }
    ],
    "one-var": [
      2,
      {
        "initialized": "never"
      }
    ],
    "operator-linebreak": [
      2,
      "after",
      {
        "overrides": {
          "?": "before",
          ":": "before"
        }
      }
    ],
    "padded-blocks": [
      2,
      "never"
    ],
    "quotes": [
      2,
      "single",
      {
        "avoidEscape": true,
        "allowTemplateLiterals": true
      }
    ],
    "rest-spread-spacing": [
      2,
      "never"
    ],
    "semi": [
      2,
      "never"
    ],
    "semi-spacing": [
      2,
      {
        "before": false,
        "after": true
      }
    ],
    "space-before-blocks": [
      2,
      "always"
    ],
    "space-before-function-paren": [
      2,
      "always"
    ],
    "space-in-parens": [
      2,
      "never"
    ],
    "space-infix-ops": 2,
    "space-unary-ops": [
      2,
      {
        "words": true,
        "nonwords": false
      }
    ],
    "spaced-comment": [
      2,
      "always",
      {
        "line": {
          "markers": [
            "*package",
            "!",
            ","
          ]
        },
        "block": {
          "balanced": true,
          "markers": [
            "*package",
            "!",
            ","
          ],
          "exceptions": [
            "*"
          ]
        }
      }
    ],
    "template-curly-spacing": [
      2,
      "never"
    ],
    "unicode-bom": [
      2,
      "never"
    ],
    "use-isnan": 2,
    "valid-typeof": 2,
    "wrap-iife": [
      2,
      "any",
      {
        "functionPrototypeMethods": true
      }
    ],
    "yield-star-spacing": [
      2,
      "both"
    ],
    "yoda": [
      2,
      "never"
    ],
    "standard/object-curly-even-spacing": [
      2,
      "either"
    ],
    "standard/array-bracket-even-spacing": [
      2,
      "either"
    ],
    "standard/computed-property-even-spacing": [
      2,
      "even"
    ],
    "promise/param-names": 2
  },
  "parserOptions": {
    "ecmaVersion": 8,
    "ecmaFeatures": {
      "globalReturn": true,
      "experimentalObjectRestSpread": true,
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "standard",
    "promise"
  ],
  "extends": "standard",
  "installedESLint": true
}
```

And just to make it clear, if I do this:
```
% cd ~/tmp/flycheck-eslint-issue
% emacs index.js
```
It just works flawlessly.

**Comment by lunaryorn** [comment]
@marc0s I think this indicates an issue in your setup; maybe your Emacs is using the wrong eslint executable when started from a different directory.

Your test repository works for with local `node_modules/` regardless from where I start it when I tell Flycheck to use the local eslint executable.

**Comment by lunaryorn** [comment]
@marc0s Can you try to reproduce the issue in [emacs -Q](http://www.lunaryorn.com/posts/reproduce-bugs-in-emacs-Q.html)?

**Comment by marc0s** [comment]
@lunaryorn shouldn't then that wrong executable's path show up in `flycheck-verify-setup` output?

Will try to use a local `node_modules` eslint

Thanks

**Comment by Simplify** [comment]
They did make some changes to ESlint's `--print-config` as of 3.8.0.

3.5.0:
```
$ eslint -h
...
  --print-config              Print the configuration to be used
```

3.9.1:
```
$eslint -f
...
  --print-config path::String  Print the configuration for the given file
```

It's worth investigating this, but I lack time until Saturday. We may need to use `flycheck-temp-file-inplace` instead of `.` in some situations.

@marc0s If your `index.js` is located in `~/tmp/flycheck-eslint-issue`, where is jour `.elsintrc.*whatever*` located?

**Comment by marc0s** [comment]
@Simplify it's located at `~/tmp/flycheck-eslint-issue/.eslintrc.yaml`

**Comment by Simplify** [comment]
And if you downgrade eslint to 3.7.1 that this problem goes away?

**Comment by marc0s** [comment]
@Simplify it doesn't :(

I have now tried with `emacs -Q` and then executed the following before opening `index.js`:

```
(add-to-list 'exec-path "/home/marcos/.nvm/v4.6.1/bin")
(package-initialize)
(global-flycheck-mode)
```
After that I `C-x f ~/tmp/flycheck-eslint-issue/index.js` and check `C-c ! v` with the same results: 
```
Syntax checkers for buffer index.js in js-mode:

  javascript-eslint (disabled)
    - may enable:  Automatically disabled!
    - executable:  Found at /home/marcos/.nvm/v4.6.1/bin/eslint
    - config file: missing

  javascript-jshint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found

  javascript-gjslint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found

  javascript-jscs (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found

  javascript-standard (disabled)
    - may enable: Automatically disabled!
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 31snapshot (package: 20161025.551)
Emacs version:    25.1.1
System:           x86_64-pc-linux-gnu
Window system:    x
```

Thanks,

**Comment by Simplify** [comment]
Well, I just updated eslint to 3.9.1 everywhere (globally and in `node_modules/.bin/` and can't reproduce your problem. I had problem that eslint returns 0 from `--print-config .`, but crashes in Flycheck because of missing `estraverse-fb` npm package (fixed with `npm install --save-dev estraverse-fb`).
This is still different from your problem. Flyckeck enables `javascript-eslint` for every file where ESlint can find valid config in my case. I tried with different projects and staring emacs from different paths, opening files in different projects directories while working on another one, everything just works fine for me.

**Comment by lunaryorn** [comment]
I can't reproduce the issue with `eslint 3.9.1` either.

If I run `emacs -Q -l ~/Developer/flycheck/test/init.el -f flycheck/init`, i.e. with the minimal testing init file from this repo, and visit an `index.js` in your testing repository it's enabled, provided that the plugins are available.

**Comment by marc0s** [comment]
Ok, I cannot reproduce the issue either if running emacs as @lunaryorn  pointed out: `emacs -Q -l ~/tmp/flycheck-eslint-issue/init.el` (no `-f flycheck/init` as it gives me a `Symbol's function definition is void: flycheck/init`).

Will try to nail down my full `init.el` to see where the problem is.

Thanks,

**Comment by lunaryorn** [comment]
@marc0s I didn't use your `init.el`, I used the one in *this* repository, under `test/init.el`.  We provide a minimal init.el file for testing.

**Comment by lunaryorn** [comment]
In any case, I'll close this issue as "invalid" then, but please do share your solution. I'm curious what the problem really is.

**Comment by marc0s** [comment]
It's a `$PATH` problem.  I was launching `emacs` through `rofi` and I'm starting my wm (i3wm) from a TTY automatically on login so probably the `$PATH` is not well set.  Launching it from an `urxvt` shell and removing any `(add-to-list 'exec-path` works now.

**Comment by Yamakaky** [comment]
I have the same problem. eslint is in node_modules/.bin/ which is not in the path. If I add it to the path before launching emacs, it is detected just fine. Question: do you know a way to auto-add it to the path? Maybe in .projectile?

**Comment by Simplify** [comment]
In your `dir-locals..el`:

```elisp
((nil . ((eval . (progn
                   (add-to-list 'exec-path (concat (locate-dominating-file default-directory ".dir-locals.el") "node_modules/.bin/")))))))
```

or use this [Emacs extension](https://github.com/codesuki/add-node-modules-path)

or in bash in `.profile` or `.bash_profile`:

```sh
export PATH="$PATH:node_modules/.bin"
```

**Comment by cpitclaudel** [comment]
(You probably want `expand-file-name` rather than `concat`)

**Comment by Yamakaky** [comment]
The last one will not do if node_modules is in a parent directory. THanks for the extension!

**Comment by chuym** [comment]
I had a problem where flycheck would report that it could not find `eslint` config file, however the problem is that the config file I have had problems (it extends a module that I don't have installed). Unfortunately, flycheck reports config errors as `config file: missing`.

Since this is the first search hit and you are also having this config file missing error, try going to the base directory of the file you are trying to get eslint enabled and run:

`eslint --print-config .`

it will output any error your config may have, fix it and flycheck will work.

https://github.com/flycheck/flycheck/blob/master/flycheck.el#L8017 - the config check for eslint just checks for the exit code, if it's different to `0` then the config file "exists".

eslint is not returning different codes for different failures though, a config error is thrown by the command line, which results in a exit code `1`, however, if the config file is missing it's handled by eslint but it _also_ returns `1`, so it's not possible currently to distinguish these errors.

In short, there isn't much flycheck can do at the moment, since eslint does not have sane exit codes and if you have problems with flycheck + eslint, make sure `eslint` works on its own first before looking into flycheck, chances are the problem is in eslint.

**Comment by cpitclaudel** [comment]
@chuym Thanks so much for documenting this here! This type of comment is invaluable :)

**Comment by heyarne** [comment]
I'm facing a similar issue and I'm not quite sure how to fix it. In my project folder:

```
$ which eslint
node_modules/.bin/eslint
$ eslint --print-config .
$ # ... prints the complete config file
$ cat .eslintrc
{
  "extends": "standard"
}
$ echo $PATH
./node_modules/.bin /Users/heyarne/.npm/bin ...
```

However, when I open spacemacs and look at my flycheck config, the local eslint is not found. It finds the one in my home folder and tells me the config file is missing or incorrect, therefore disabling `javascript-eslint`:

```
Syntax checkers for buffer index.js in js2-mode:

  javascript-eslint (disabled)
    - may enable:  Automatically disabled!
    - executable:  Found at /Users/heyarne/.npm/bin/eslint
    - config file: missing or incorrect

  javascript-jshint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found

  javascript-jscs (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found

  javascript-standard
    - may enable: yes
    - executable: Found at /Users/heyarne/.npm/bin/standard

  drupal-phpcs (disabled)
    - may enable: Automatically disabled!
    - predicate:  nil
    - executable: Not found

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20171109.216)
Emacs version:    25.3.1
System:           x86_64-apple-darwin16.7.0
Window system:    ns
```

All the necessary npm modules are installed. `eslint .` Gives me the correct linter output. I'm using fish 2.7.1 as my shell. Any idea what's wrong?

**Comment by Simplify** [comment]
@heyarne , are you on macOS? Emacs there does not take `$PATH` from shell. We also had similar problems on GNU/Linux. Try [using this](https://github.com/purcell/exec-path-from-shell).

For example i have following code in my init files:
```elisp
(when (memq window-system '(mac ns))                                                                                                                      
    (setq exec-path-from-shell-variables '("PATH" "MANPATH" "SSH_CLIENT" "HOSTNAME"                                                                         
                                           "GTAGSCONF" "GTAGSLABEL" "RUST_SRC_PATH"                                                                         
                                           "HISTFILE" "HOME" "GOPATH" "GOROOT" "GOEXEC"))                                                                   
    (exec-path-from-shell-initialize)) 
```

**Comment by tordans** [comment]
_In case anyone googles this…_

I had to run (from my projects' directory) `eslint --print-config .eslintrc` (where `.eslintrc` is the name of my config file). It showed me an error in my config.
